File: backend/app/api/v1/endpoints/kb.py
"""Knowledge-base endpoints (P3-2): articles/versions/rollback, categories, search.

Static segments register before `{id}` param segments. Feature-flag/permission
guards live in the service (routes stay registered so openapi keeps every key).
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Query

from app.api.deps import DbDep, UserDep
from app.core.response import Result
from app import schemas
from app.services import kb_service

router = APIRouter(prefix="/kb", tags=["kb"])


@router.get("/articles", response_model=Result)
def list_articles(
    db: DbDep,
    user: UserDep,
    keyword: str | None = None,
    category_id: int | None = None,
    tag: str | None = None,
    visibility: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 10,
):
    filters = {"keyword": keyword, "category_id": category_id, "tag": tag,
               "visibility": visibility}
    return Result.ok(kb_service.list_articles(db, user, filters, page, size))


@router.post("/articles", response_model=Result)
def create_article(db: DbDep, user: UserDep, data: schemas.KbArticleCreate):
    return Result.ok(kb_service.create_article(db, user, data))


@router.get("/articles/{article_id}", response_model=Result)
def get_article(db: DbDep, user: UserDep, article_id: int):
    return Result.ok(kb_service.get_article(db, user, article_id))


@router.put("/articles/{article_id}", response_model=Result)
def update_article(db: DbDep, user: UserDep, article_id: int, data: schemas.KbArticleUpdate):
    return Result.ok(kb_service.update_article(db, user, article_id, data))


@router.delete("/articles/{article_id}", response_model=Result)
def delete_article(db: DbDep, user: UserDep, article_id: int):
    return Result.ok(kb_service.delete_article(db, user, article_id))


@router.get("/articles/{article_id}/versions", response_model=Result)
def list_versions(db: DbDep, user: UserDep, article_id: int):
    return Result.ok(kb_service.list_versions(db, user, article_id))


@router.get("/articles/{article_id}/versions/{version}", response_model=Result)
def get_version(db: DbDep, user: UserDep, article_id: int, version: int):
    return Result.ok(kb_service.get_version(db, user, article_id, version))


@router.post("/articles/{article_id}/rollback", response_model=Result)
def rollback_article(db: DbDep, user: UserDep, article_id: int, data: schemas.KbRollbackIn):
    return Result.ok(kb_service.rollback_article(db, user, article_id, data))


@router.get("/categories", response_model=Result)
def list_categories(db: DbDep, user: UserDep):
    return Result.ok(kb_service.list_categories(db, user))


@router.post("/categories", response_model=Result)
def create_category(db: DbDep, user: UserDep, data: schemas.KbCategoryCreate):
    return Result.ok(kb_service.create_category(db, user, data))


@router.put("/categories/{category_id}", response_model=Result)
def update_category(db: DbDep, user: UserDep, category_id: int, data: schemas.KbCategoryUpdate):
    return Result.ok(kb_service.update_category(db, user, category_id, data))


@router.delete("/categories/{category_id}", response_model=Result)
def delete_category(db: DbDep, user: UserDep, category_id: int):
    return Result.ok(kb_service.delete_category(db, user, category_id))


@router.get("/search", response_model=Result)
def search_articles(
    db: DbDep,
    user: UserDep,
    q: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 10,
):
    return Result.ok(kb_service.search_articles(db, user, q, page, size))

File: frontend/src/api/kb.ts
/**
 * Knowledge-base API (P3-2) — frontend/src/api/kb.ts
 * Contract: api-design-v3.md §6 (13 endpoints). Shapes @ backend/app/schemas/kb.py.
 * The service gates on `feature.kb` (default False -> HTTP 400 "feature disabled").
 */
import http from './http'
import type {
  Page,
  Result,
  KbArticleOut,
  KbArticleDetail,
  KbArticleQuery,
  KbArticleCreate,
  KbArticleUpdate,
  KbVersionListOut,
  KbVersionOut,
  KbCategoryListOut,
  KbCategoryOut,
  KbCategoryCreate,
  KbCategoryUpdate,
} from './types'

// ---------------------------------------------------------------- articles

export async function listArticles(params?: KbArticleQuery): Promise<Page<KbArticleOut>> {
  const { data } = await http.get<Result<Page<KbArticleOut>>>('/kb/articles', { params })
  return data.data
}

export async function getArticle(id: number): Promise<KbArticleDetail> {
  const { data } = await http.get<Result<KbArticleDetail>>(`/kb/articles/${id}`)
  return data.data
}

export async function createArticle(payload: KbArticleCreate): Promise<KbArticleOut> {
  const { data } = await http.post<Result<KbArticleOut>>('/kb/articles', payload)
  return data.data
}

export async function updateArticle(id: number, payload: KbArticleUpdate): Promise<KbArticleOut> {
  const { data } = await http.put<Result<KbArticleOut>>(`/kb/articles/${id}`, payload)
  return data.data
}

export async function deleteArticle(id: number): Promise<void> {
  await http.delete<Result<unknown>>(`/kb/articles/${id}`)
}

export async function listVersions(id: number): Promise<KbVersionListOut> {
  const { data } = await http.get<Result<KbVersionListOut>>(`/kb/articles/${id}/versions`)
  return data.data
}

export async function getVersion(id: number, version: number): Promise<KbVersionOut> {
  const { data } = await http.get<Result<KbVersionOut>>(`/kb/articles/${id}/versions/${version}`)
  return data.data
}

export async function rollbackArticle(id: number, version: number): Promise<KbArticleOut> {
  const { data } = await http.post<Result<KbArticleOut>>(`/kb/articles/${id}/rollback`, { version })
  return data.data
}

// ---------------------------------------------------------------- categories

export async function listCategories(): Promise<KbCategoryListOut> {
  const { data } = await http.get<Result<KbCategoryListOut>>('/kb/categories')
  return data.data
}

export async function createCategory(payload: KbCategoryCreate): Promise<KbCategoryOut> {
  const { data } = await http.post<Result<KbCategoryOut>>('/kb/categories', payload)
  return data.data
}

export async function updateCategory(id: number, payload: KbCategoryUpdate): Promise<KbCategoryOut> {
  const { data } = await http.put<Result<KbCategoryOut>>(`/kb/categories/${id}`, payload)
  return data.data
}

export async function deleteCategory(id: number): Promise<void> {
  await http.delete<Result<unknown>>(`/kb/categories/${id}`)
}

// ---------------------------------------------------------------- search

export async function searchArticles(
  q: string,
  params?: { page?: number; size?: number },
): Promise<Page<KbArticleOut>> {
  const { data } = await http.get<Result<Page<KbArticleOut>>>('/kb/search', {
    params: { q, ...params },
  })
  return data.data
}
